man/cafe-network-properties.1
.\" Copyright (C) 2014 Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\"You should have received a copy of the GNU General Public License along
.\"with this program; if not, write to the Free Software Foundation, Inc.,
.\"51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.TH cafe-network-properties 1 "2014\-05\-02" "CAFE"
.SH NAME
cafe-network-properties \- configure CAFE network proxy
.SH SYNOPSIS
.B cafe-network-properties
.RI [ OPTIONS... ]
.SH DESCRIPTION
This executable is part of the package 'cafe\-control\-center': The CAFE Control Center.
.PP
The CAFE network proxy preference tool enables you to configure how
your system connects to networks. You can configure the desktop to
connect to a proxy server, and specify the details of the proxy server.
.PP
A proxy server is a server that intercepts requests to another server,
and fulfills the request itself, if it can. You can enter the Domain
Name Service (DNS) name or the Internet Protocol (IP) address of the
proxy server.
.SH OPTIONS
This program accepts the standard CAFE and CTK options.
.SH AUTHORS
.B cafe-network-properties
This manual page was writtenby Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
for the Debian project (but may be used by others).
.SH SEE ALSO
Latest version of the CAFE Desktop User Guide for your platform.
.PP
.BR "cafe-control-center" (1),
.BR "ctk-options" (7)


NEWS
### cafe-control-center 1.24.1

  * Translations update
  * cafe-theme-info: Use of memory after it is freed
  * build: silent build warnings for distcheck
  * use default CAFE_COMPILE_WARNINGS level
  * Port libslab_handle_g_error to the built-in GLib logging framework
  * cafe-about-me: Fix memory leak
  * appearance: Fix memory leak
  * Do not collect the translation for Icon

### cafe-control-center 1.24.0

  * Translations update
  * cafe-theme-info: fix memory leak
  * cafe-wp-info: fix memory leak
  * glib-genmarshal: Using --header and --body at the same time is deprecated
  * build: show compiler/flags info on configure summary
  * build: do not override WARN_CFLAGS
  * build: use warning flags
  * time-admin: make portable backward file path
  * appearance: backgrounds dir should be portable
  * appearance: remove I/O error on window close
  * autoconf: AC_PROG_CC was called before AX_CHECK_ENABLE_DEBUG
  * Fix build using gcc 10 -fno-common flag

### cafe-control-center 1.23.3

  * Translations update
  * Update copyright to 2020
  * about-me: Fix UI validation
  * mouse: Fix UI validation
  * time-admin: Add license info in time-admin.ui file
  * time-admin: Fix UI validation
  * at-properties: Fix UI validation
  * Search system themes in system data dirs
  * display: Fix UI validation
  * display: Enable help button
  * windows: Add window-properties.ui
  * avoid redundant redeclarations
  * keybindings: Fix schema creation in makefile
  * appearance: Fix desktop file creation
  * shell: Fix typo in keyword name
  * migrate from intltool to gettext
  * default-applications: Load icons as cairo surfaces
  * icons: Force uniform size on icons (#505)
  * remove warning: "__sun" is not defined, evaluates to 0
  * fv: thumbnailer: add font/collection to the supported mimetypes
  * font-viewer: Add font/ttf and font/otf mime types
  * time-admin: fix memory leak
  * time-admin: Don't share «void QuitApp(TimeAdmin *ta)» method
  * default-applications: Remove messenger options from ui file
  * time-admin: Remove unreachable code
  * time-admin: Simplify date & time displaying
  * time-admin: Save button cannot be enabled if NTP sync is active
  * time-admin: Display current time in time-admin.ui on startup
  * time-admin: Rename adjustament to adjustment in time-admin.ui
  * time-admin: A11y time-admin.ui
  * time-admin: Add time-admin.ui
  * default-applications: Fix Word Processor top margin

### cafe-control-center 1.23.2

  * Translations update
  * travis: update cafe-desktop and cafe-settings-daemon required versions
  * mouse: Add support for Acceleration Profiles
  * appearance: fix memory leak
  * appearance: add new font for user
  * Stop propagate error
  * at-properties: migrate from dbus-glib to gdbus

### cafe-control-center 1.23.1

  * Translations update
  * fix build error when not installed gla11y
  * remove lightdm-settings support
  * launch lightdm-settings if installed
  * use current running DM preferences program
  * support lightdm-ctk-greeter-settings as login preferences program
  * replace Tab with space
  * drop libXxf86 dependency
  * Use python3 for gla11y
  * Add scroll tabs with mouse wheel in severals capplets
  * libslab: bookmark-agent.c avoid deprecated g_type_class_add_private
  * font-viewer: avoid deprecated 'g_type_class_add_private'
  * Remove compilation warnings
  * time-admin: Don't set window title for message-box dialog
  * time-admin: Don't set font size/color for regular controls
  * use theme's icon preferences-system-time
  * time-admin: Fix license notice in time-zone.h|c, time-map.h|c
  * capplets: Remove the macro CAFELOCALEDIR where it's not required
  * capplets: Use capplet_init
  * time-admin: shortcuts, images & shared code for buttons
  * common: Remove capplet-stock-icons
  * time-admin: Remove po-timezones
  * Remove trailing whitespaces

### cafe-control-center 1.23.0

  * Translations update
  * bump require of m-s-d and m-d
  * Add orca to Visual combo box if available
  * Add support for checking .ui accessibility with gla11y
  * Do not reload NTP and modify the button arrangement when starting the service
  * Authentication to add modification time
  * Use default color to set label and remove the maximum button
  * use Makefile pkgdatadir
  * g_date_time_unref datetime
  * Display the correct time when selecting the time zone
  * Memory leak: g_error_free(error);
  * Modify cafe-time-admin.desktop.in file format
  * enable to build time-admin for cafe-control-center
  * add time-admin source code
  * add timezone translation files for time-admin
  * add image resources for time-admin applet
  * Populate initial messenger list with irc scheme handler apps
  * Set default icq/irc/ircs/sip/xmpp scheme handler
  * Add instant messenger keybinding into cafe-keybinding-properties
  * Change the icon for instant messaging application
  * default-applications: Add a widget for a default messenger application
  * appearance.ui (cursor size scale)
  * libslab: avoid deprecated g_type_class_add_private
  * Create FUNDING.yml
  * typing-break: avoid deprecated 'g_type_class_add_private'
  * file-transfer-dialog.c: avoid deprecated 'g_type_class_add_private'
  * Fix the icon sizes in Preferred Applications selector
  * capplets: Use gresources for UI files
  * [ci] Add cppcheck html report
  * Fix grammar mistake
  * Migrate from dbus-glib to GDBus
  * Use system data dirs to locate key bindings
  * control-center: Fix memory leak
  * Use CtkSeparatorMenuItem in Apparence preview
  * Remove dead store detected by clang
  * default-applications: Improve UI accessibility
  * Use keyboard device icon like in other hardware devices
  * Use icons provided by themes and remove not used icons
  * help: update copyright
  * [ci] Enable Clang Static Analyzer
  * add installation instructions

### cafe-control-center 1.22.0

  * Translations update
  * Initialize travis support

capplets/network/cafe-network-properties.c
/* cafe-network-properties.c: network preferences capplet
 *
 * Copyright (C) 2002 Sun Microsystems Inc.
 *               2012 Stefano Karapetsas
 *
 * Written by: Mark McLoughlin <mark@skynet.ie>
 *             Stefano Karapetsas <stefano@karapetsas.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <stdlib.h>
#include <string.h>
#include <gio/gio.h>
#include <glib/gi18n.h>

#include "capplet-util.h"

/* With GSettings migration, CAFE is going to share proxy settings with
 * GNOME3 (in gsettings-desktop-schemas package).
 * GSettings schemas is http://git.gnome.org/browse/gsettings-desktop-schemas/tree/
 * schemas/org.gnome.system.proxy.gschema.xml.in.in
 * Sharing the same schema, we lose the locations feature.
 */

/*
 * use enum from gsettings-desktop-schema
 * http://git.gnome.org/browse/gsettings-desktop-schemas/tree/headers/gdesktop-enums.h#n26
 */
enum GDesktopProxyMode
{
	G_DESKTOP_PROXY_MODE_NONE,
	G_DESKTOP_PROXY_MODE_MANUAL,
	G_DESKTOP_PROXY_MODE_AUTO
};

enum {
	COL_NAME,
	COL_STYLE
};

#define PROXY_SCHEMA              "org.gnome.system.proxy"
#define PROXY_MODE_KEY            "mode"
#define PROXY_AUTOCONFIG_URL_KEY  "autoconfig-url"
#define IGNORE_HOSTS_KEY          "ignore-hosts"

#define HTTP_PROXY_SCHEMA         "org.gnome.system.proxy.http"
#define HTTP_PROXY_HOST_KEY       "host"
#define HTTP_PROXY_PORT_KEY       "port"
#define HTTP_USE_AUTH_KEY         "use-authentication"
#define HTTP_AUTH_USER_KEY        "authentication-user"
#define HTTP_AUTH_PASSWD_KEY      "authentication-password"

#define HTTPS_PROXY_SCHEMA        "org.gnome.system.proxy.https"
#define SECURE_PROXY_HOST_KEY     "host"
#define SECURE_PROXY_PORT_KEY     "port"

#define FTP_PROXY_SCHEMA          "org.gnome.system.proxy.ftp"
#define FTP_PROXY_HOST_KEY        "host"
#define FTP_PROXY_PORT_KEY        "port"

#define SOCKS_PROXY_SCHEMA        "org.gnome.system.proxy.socks"
#define SOCKS_PROXY_HOST_KEY      "host"
#define SOCKS_PROXY_PORT_KEY      "port"

static CtkWidget *details_dialog = NULL;
static GSList *ignore_hosts = NULL;
static CtkTreeModel *model = NULL;

static GSettings *proxy_settings = NULL;
static GSettings *http_proxy_settings = NULL;
static GSettings *https_proxy_settings = NULL;
static GSettings *ftp_proxy_settings = NULL;
static GSettings *socks_proxy_settings = NULL;

static CtkTreeModel *
create_listmodel(void)
{
	CtkListStore *store;

	store = ctk_list_store_new(1, G_TYPE_STRING);

	return CTK_TREE_MODEL(store);
}

static CtkTreeModel *
populate_listmodel(CtkListStore *store, GSList *list)
{
	CtkTreeIter iter;
	GSList *pointer;

	ctk_list_store_clear(store);

	pointer = list;
	while(pointer)
	{
		ctk_list_store_append(store, &iter);
		ctk_list_store_set(store, &iter, 0, (char *) pointer->data, -1);
		pointer = g_slist_next(pointer);
	}

	return CTK_TREE_MODEL(store);
}

static CtkWidget *
config_treeview(CtkTreeView *tree, CtkTreeModel *model)
{
	CtkCellRenderer *renderer;

	renderer = ctk_cell_renderer_text_new();
	ctk_tree_view_insert_column_with_attributes(CTK_TREE_VIEW(tree),
												-1, "Hosts", renderer,
												"text", 0, NULL);

	ctk_tree_view_set_model(CTK_TREE_VIEW(tree), model);

	return CTK_WIDGET(tree);
}

static CtkWidget*
_ctk_builder_get_widget (CtkBuilder *builder, const gchar *name)
{
	return CTK_WIDGET (ctk_builder_get_object (builder, name));
}

static void
read_ignore_hosts_from_gsettings (void)
{
	gchar **array;
	ignore_hosts = NULL;
	gint i;
	array = g_settings_get_strv (proxy_settings, IGNORE_HOSTS_KEY);
	if (array != NULL)
	{
		for (i = 0; array[i]; i++)
		{
			ignore_hosts = g_slist_append (ignore_hosts, g_strdup (array[i]));
		}
	}
	g_strfreev (array);
}

static void
save_ignore_hosts_to_gsettings (void)
{
	GArray *array;
	GSList *l;
	array = g_array_new (TRUE, TRUE, sizeof (gchar *));
	for (l = ignore_hosts; l; l = l->next) {
		array = g_array_append_val (array, l->data);
	}
	g_settings_set_strv (proxy_settings, IGNORE_HOSTS_KEY, (const gchar **) array->data);
	g_array_free (array, TRUE);
}

static void
cb_add_url (CtkButton *button, gpointer data)
{
	CtkBuilder *builder = CTK_BUILDER (data);

	const gchar *entry_text = ctk_entry_get_text (CTK_ENTRY (ctk_builder_get_object (builder, "entry_url")));
	if (entry_text == NULL || strlen (entry_text) == 0) {
		return;
	}

	ignore_hosts = g_slist_append(ignore_hosts, g_strdup (entry_text));
	populate_listmodel(CTK_LIST_STORE(model), ignore_hosts);
	ctk_entry_set_text(CTK_ENTRY (ctk_builder_get_object (builder,
							     "entry_url")), "");

	save_ignore_hosts_to_gsettings ();
}

static void
cb_remove_url (CtkButton *button, gpointer data)
{
	CtkBuilder *builder = CTK_BUILDER (data);
	CtkTreeSelection *selection;
	CtkTreeIter       iter;

	selection = ctk_tree_view_get_selection (CTK_TREE_VIEW (ctk_builder_get_object (builder, "treeview_ignore_host")));
	if (ctk_tree_selection_get_selected(selection, &model, &iter))
	{
		gchar *url;
		GSList *pointer;

		ctk_tree_model_get (model, &iter, 0, &url, -1);

		pointer = ignore_hosts;
		while(pointer)
		{
			if(strcmp(url, (char *) pointer->data) == 0)
			{
				g_free (pointer->data);
				ignore_hosts = g_slist_delete_link(ignore_hosts, pointer);
				break;
			}
			pointer = g_slist_next(pointer);
		}

		g_free(url);
		populate_listmodel(CTK_LIST_STORE(model), ignore_hosts);

		save_ignore_hosts_to_gsettings ();
	}
}

static void
cb_dialog_response (CtkDialog *dialog, gint response_id)
{
	if (response_id == CTK_RESPONSE_HELP)
		capplet_help (CTK_WINDOW (dialog),
			"goscustdesk-50");
	else if (response_id == CTK_RESPONSE_CLOSE || response_id == CTK_RESPONSE_DELETE_EVENT)
	{
		if (ignore_hosts) {
			g_slist_foreach (ignore_hosts, (GFunc) g_free, NULL);
			g_slist_free (ignore_hosts);
		}

		ctk_main_quit ();
	}
}

static void
cb_details_dialog_response (CtkDialog *dialog, gint response_id)
{
	if (response_id == CTK_RESPONSE_HELP)
		capplet_help (CTK_WINDOW (dialog),
			      "goscustdesk-50");
	else {
		ctk_widget_destroy (CTK_WIDGET (dialog));
		details_dialog = NULL;
	}
}

static void
cb_use_auth_toggled (CtkToggleButton *toggle,
		     CtkWidget *table)
{
	ctk_widget_set_sensitive (table, ctk_toggle_button_get_active (toggle));
}

static void
cb_http_details_button_clicked (CtkWidget *button,
			        CtkWidget *parent)
{
	CtkBuilder *builder;
	gchar *builder_widgets[] = { "details_dialog", NULL };
	GError *error = NULL;
	CtkWidget *widget;

	if (details_dialog != NULL) {
		ctk_window_present (CTK_WINDOW (details_dialog));
		ctk_widget_grab_focus (details_dialog);
		return;
	}

	builder = ctk_builder_new ();
	if (ctk_builder_add_objects_from_resource (builder, "/org/cafe/mcc/network/cafe-network-properties.ui",
					           builder_widgets, &error) == 0) {
		g_warning ("Could not load details dialog: %s", error->message);
		g_error_free (error);
		g_object_unref (builder);
		return;
	}

	details_dialog = widget = _ctk_builder_get_widget (builder,
							   "details_dialog");

	ctk_window_set_transient_for (CTK_WINDOW (widget), CTK_WINDOW (parent));

	g_signal_connect (ctk_builder_get_object (builder, "use_auth_checkbutton"),
			  "toggled",
			  G_CALLBACK (cb_use_auth_toggled),
			  _ctk_builder_get_widget (builder, "auth_table"));

	g_settings_bind (http_proxy_settings, HTTP_USE_AUTH_KEY,
			ctk_builder_get_object (builder, "use_auth_checkbutton"), "active",
			G_SETTINGS_BIND_DEFAULT);
	g_settings_bind (http_proxy_settings, HTTP_AUTH_USER_KEY,
			ctk_builder_get_object (builder, "username_entry"), "text",
			G_SETTINGS_BIND_DEFAULT);
	g_settings_bind (http_proxy_settings, HTTP_AUTH_PASSWD_KEY,
			ctk_builder_get_object (builder, "password_entry"), "text",
			G_SETTINGS_BIND_DEFAULT);

	g_signal_connect (widget, "response",
			  G_CALLBACK (cb_details_dialog_response), NULL);

	capplet_set_icon (widget, "network-server");

	ctk_widget_show_all (widget);
}

static void
proxy_mode_gsettings_changed (GSettings *settings,
				   gchar *key,
				   CtkBuilder *builder)
{
	int mode;
	mode = g_settings_get_enum (settings, PROXY_MODE_KEY);
	if (mode == G_DESKTOP_PROXY_MODE_NONE)
		ctk_toggle_button_set_active (CTK_TOGGLE_BUTTON (ctk_builder_get_object(builder, "none_radiobutton")), TRUE);
	else if (mode == G_DESKTOP_PROXY_MODE_AUTO)
		ctk_toggle_button_set_active (CTK_TOGGLE_BUTTON (ctk_builder_get_object(builder, "auto_radiobutton")), TRUE);
	else if (mode == G_DESKTOP_PROXY_MODE_MANUAL)
		ctk_toggle_button_set_active (CTK_TOGGLE_BUTTON (ctk_builder_get_object(builder, "manual_radiobutton")), TRUE);
	ctk_widget_set_sensitive (_ctk_builder_get_widget (builder, "manual_box"),
				  mode == G_DESKTOP_PROXY_MODE_MANUAL);
	ctk_widget_set_sensitive (_ctk_builder_get_widget (builder, "auto_box"),
				  mode == G_DESKTOP_PROXY_MODE_AUTO);
}

static void
proxy_mode_radiobutton_clicked_cb (CtkWidget *widget,
				   CtkBuilder *builder)
{
	GSList *mode_group;
	int mode;
	int old_mode;

	if (!ctk_toggle_button_get_active (CTK_TOGGLE_BUTTON(widget)))
		return;

	mode_group = g_slist_copy (ctk_radio_button_get_group
		(CTK_RADIO_BUTTON (ctk_builder_get_object (builder, "none_radiobutton"))));
	mode_group = g_slist_reverse (mode_group);
	mode = g_slist_index (mode_group, widget);
	g_slist_free (mode_group);

	old_mode = g_settings_get_enum (proxy_settings, PROXY_MODE_KEY);

	if (mode == old_mode)
		return;

	g_settings_set_enum (proxy_settings, PROXY_MODE_KEY, mode);
}

static void
connect_sensitivity_signals (CtkBuilder *builder, GSList *mode_group)
{
	for (; mode_group != NULL; mode_group = mode_group->next)
	{
		g_signal_connect (G_OBJECT (mode_group->data), "clicked",
				  G_CALLBACK(proxy_mode_radiobutton_clicked_cb),
				  builder);
	}
}

static void
cb_ignore_hosts_gsettings_changed (GSettings *settings, gchar *key, gpointer user_data)
{
	g_slist_foreach (ignore_hosts, (GFunc) g_free, NULL);
	g_slist_free (ignore_hosts);

	read_ignore_hosts_from_gsettings ();

	populate_listmodel (CTK_LIST_STORE (model), ignore_hosts);
}

static void
setup_dialog (CtkBuilder *builder)
{
	GSList *mode_group;

	g_signal_connect (proxy_settings, "changed::" IGNORE_HOSTS_KEY,
			G_CALLBACK (cb_ignore_hosts_gsettings_changed), NULL);

	/* Mode */
	proxy_mode_gsettings_changed (proxy_settings, PROXY_MODE_KEY, builder);
	mode_group = ctk_radio_button_get_group (CTK_RADIO_BUTTON (ctk_builder_get_object (builder, "none_radiobutton")));
	connect_sensitivity_signals (builder, mode_group);
	g_signal_connect (proxy_settings, "changed::" PROXY_MODE_KEY,
					  G_CALLBACK (proxy_mode_gsettings_changed), builder);

	/* Http */
	g_settings_bind (http_proxy_settings, HTTP_PROXY_PORT_KEY,
			ctk_builder_get_object (builder, "http_port_spinbutton"), "value",
			G_SETTINGS_BIND_DEFAULT);
	g_settings_bind (http_proxy_settings, HTTP_PROXY_HOST_KEY,
			ctk_builder_get_object (builder, "http_host_entry"), "text",
			G_SETTINGS_BIND_DEFAULT);
	g_signal_connect (ctk_builder_get_object (builder, "details_button"),
			  "clicked",
			  G_CALLBACK (cb_http_details_button_clicked),
			  _ctk_builder_get_widget (builder, "network_dialog"));

	/* Secure */
	g_settings_bind (https_proxy_settings, SECURE_PROXY_PORT_KEY,
			ctk_builder_get_object (builder, "secure_port_spinbutton"), "value",
			G_SETTINGS_BIND_DEFAULT);
	g_settings_bind (https_proxy_settings, SECURE_PROXY_HOST_KEY,
			ctk_builder_get_object (builder, "secure_host_entry"), "text",
			G_SETTINGS_BIND_DEFAULT);

	/* Ftp */
	g_settings_bind (ftp_proxy_settings, FTP_PROXY_PORT_KEY,
			ctk_builder_get_object (builder, "ftp_port_spinbutton"), "value",
			G_SETTINGS_BIND_DEFAULT);
	g_settings_bind (ftp_proxy_settings, FTP_PROXY_HOST_KEY,
			ctk_builder_get_object (builder, "ftp_host_entry"), "text",
			G_SETTINGS_BIND_DEFAULT);

	/* Socks */
	g_settings_bind (socks_proxy_settings, SOCKS_PROXY_PORT_KEY,
			ctk_builder_get_object (builder, "socks_port_spinbutton"), "value",
			G_SETTINGS_BIND_DEFAULT);
	g_settings_bind (socks_proxy_settings, SOCKS_PROXY_HOST_KEY,
			ctk_builder_get_object (builder, "socks_host_entry"), "text",
			G_SETTINGS_BIND_DEFAULT);

	/* Autoconfiguration */
	g_settings_bind (proxy_settings, PROXY_AUTOCONFIG_URL_KEY,
			ctk_builder_get_object (builder, "autoconfig_entry"), "text",
			G_SETTINGS_BIND_DEFAULT);

	g_signal_connect (ctk_builder_get_object (builder, "network_dialog"),
			"response", G_CALLBACK (cb_dialog_response), NULL);

	read_ignore_hosts_from_gsettings ();

	model = create_listmodel();
	populate_listmodel(CTK_LIST_STORE(model), ignore_hosts);
	config_treeview(CTK_TREE_VIEW(ctk_builder_get_object (builder, "treeview_ignore_host")), model);

	g_signal_connect (ctk_builder_get_object (builder, "button_add_url"),
			  "clicked", G_CALLBACK (cb_add_url), builder);
	g_signal_connect (ctk_builder_get_object (builder, "entry_url"),
			  "activate", G_CALLBACK (cb_add_url), builder);
	g_signal_connect (ctk_builder_get_object (builder, "button_remove_url"),
			  "clicked", G_CALLBACK (cb_remove_url), builder);
}

int
main (int argc, char **argv)
{
	CtkBuilder  *builder;
	GError *error = NULL;
	gchar *builder_widgets[] = {"network_dialog", "adjustment1",
				    "adjustment2", "adjustment3", "adjustment4",
				    "delete_button_img", NULL};
	CtkWidget   *widget;

	capplet_init (NULL, &argc, &argv);

	builder = ctk_builder_new ();
	if (ctk_builder_add_objects_from_resource (builder, "/org/cafe/mcc/network/cafe-network-properties.ui",
					           builder_widgets, &error) == 0) {
		g_warning ("Could not load main dialog: %s",
			   error->message);
		g_error_free (error);
		g_object_unref (builder);
		return (EXIT_FAILURE);
	}

	proxy_settings = g_settings_new (PROXY_SCHEMA);
	http_proxy_settings = g_settings_new (HTTP_PROXY_SCHEMA);
	https_proxy_settings = g_settings_new (HTTPS_PROXY_SCHEMA);
	ftp_proxy_settings = g_settings_new (FTP_PROXY_SCHEMA);
	socks_proxy_settings = g_settings_new (SOCKS_PROXY_SCHEMA);

	setup_dialog (builder);
	widget = _ctk_builder_get_widget (builder, "network_dialog");

        CtkNotebook* nb = CTK_NOTEBOOK (_ctk_builder_get_widget (builder, "notebook1"));
        ctk_widget_add_events (CTK_WIDGET (nb), CDK_SCROLL_MASK);
        g_signal_connect (CTK_WIDGET (nb), "scroll-event",
                          G_CALLBACK (capplet_dialog_page_scroll_event_cb),
                          CTK_WINDOW (widget));

	capplet_set_icon (widget, "network-server");
	ctk_widget_show_all (widget);
	ctk_main ();

	g_object_unref (builder);
	g_object_unref (proxy_settings);
	g_object_unref (http_proxy_settings);
	g_object_unref (https_proxy_settings);
	g_object_unref (ftp_proxy_settings);
	g_object_unref (socks_proxy_settings);

	return 0;
}
